### PR TITLE
handle pure callsites as special callinfo

### PR DESCRIPTION
**Pull request by aviatesk**
Follows up #180.

@timholy What do you think on this ? I don't think it hurts if we can't recurse into a pure callsite (just because inference doesn't really recurse into it)

### CODE REVIEW

File: src/reflection.jl [hunk 1]
```diff
@@ -51,6 +51,15 @@ function find_callsites(interp::CthulhuInterpreter, CI::Union{Core.CodeInfo, IRC
...
                 rt = argextype(SSAValue(id), CI, sptypes, slottypes)
+                # in unoptimized IR, there may be `slot = rhs` expressions, which `argextype` doesn't handle
+                # so extract rhs for such an case
+                args = c.args
+                if !optimize
+                    if isexpr(c, :(=))
+                        args = c.args[2].args
+                    end
+                end
+                types = mapany(arg -> widenconst(argextype(arg, CI, sptypes, slottypes)), args)
```
Comment (timholy) [inline src/reflection.jl:62]: Very good to move this, I was curious about its former location.

File: test/runtests.jl [hunk 2]
```diff
@@ -63,9 +63,15 @@ isordered(::Type{T}) where {T<:AbstractDict} = false
     @test isempty(callsites)
 
     callsites = find_callsites_by_ftt(iterate, Tuple{SVector{3,Int}, Tuple{SOneTo{3}}}; optimize=false)
-    @test occursin("< pure, uncached >", string(callsites[1]))
+    @test occursin("< pure >", string(callsites[1]))
     @test occursin(r"< constprop > getindex\(::.*Const.*,::.*Const\(1\)\)::.*Const\(1\)", string(callsites[2]))
 
+    # handle pure
```
Comment (timholy) [inline test/runtests.jl:69]: Move this comment up above line 65?

**Review by timholy — APPROVED** [review]
This looks great and I approve it to replace #180.

File: src/callsite.jl [hunk 3]
```diff
@@ -207,6 +210,13 @@ function show_callinfo(limiter, ci::Union{MultiCallInfo, FailedCallInfo, Generat
     __show_limited(limiter, name, tt, rt)
 end
 
+function show_callinfo(limiter, (; argtypes, rt)::PureCallInfo)
```
[the remark this answers is not in the document]
Comment (aviatesk) [reply src/callsite.jl:213]: Yeah, this syntax fits into my coding style very quickly. Really appreciated, @simeonschaub !

**Pull request merged**